packages/shadydom/src/patches/ParentNode.js

```diff
@@ -128,24 +131,248 @@ export const ParentNodePatches = utils.getOwnPropertyDescriptors({
   },
 });
 
+/**
+ * Performs the equivalent of `querySelectorAll` within Shady DOM's logical
+ * model of the tree for a selector list.
+ *
+ * See <./logicalQuerySelectorAll.md> for implementation details.
+ *
+ * @param {!Element} contextElement
+ * @param {string} selectorList
+ * @return {!Array<!Element>}
+ */
+const logicalQuerySelectorAll = (contextElement, selectorList) => {
+  /**
+   * A key-renamed version of the return value of `extractSelectors`, which
+   * describes a single complex selector.
+   *
+   * @typedef {{
+   *   compoundSelectors: !Array<string>,
+   *   combinators: !Array<string>,
+   * }}
+   */
+  let ComplexSelectorParts;
+
+  /**
+   * @type {!Array<!ComplexSelectorParts>}
+   */
+  const complexSelectors = extractSelectors(selectorList)
+    .map((complexSelector) => {
+      const {
+        'selectors': compoundSelectors,
+        'joiners': combinators,
+      } = splitSelectorBlocks(complexSelector);
+
+      return {
+        compoundSelectors,
+        combinators,
+      };
+    })
+    .filter(({compoundSelectors}) => compoundSelectors.length > 0);
```

Comment:
For what case is this filter needed?

Reply:
This probably isn't necessary. I just haven't tested how the library responds to all whitespace / the empty string yet.

Reply:
Removed.